File: packages/wasmvm/wasmclient/ts/wasmclient/isc/offledgerrequest.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as isc from "./index";
import * as wasmlib from "wasmlib"

export class OffLedgerSignatureScheme {
    keyPair: isc.KeyPair;
    signature: u8[] = [];

    public constructor(keyPair: isc.KeyPair) {
        this.keyPair = keyPair;
    }
}

export class OffLedgerRequest {
    chainID: wasmlib.ScChainID;
    contract: wasmlib.ScHname;
    entryPoint: wasmlib.ScHname;
    params: u8[];
    signatureScheme: isc.OffLedgerSignatureScheme = new isc.OffLedgerSignatureScheme(new isc.KeyPair([]));
    nonce: u64;
    allowance: wasmlib.ScAssets = new wasmlib.ScAssets([]);
    gasBudget: u64 = 0;

    public constructor(chainID: wasmlib.ScChainID, contract: wasmlib.ScHname, entryPoint: wasmlib.ScHname, params: u8[], nonce: u64) {
        this.chainID = chainID;
        this.contract = contract;
        this.entryPoint = entryPoint;
        this.params = params;
        this.nonce = nonce;
    }

    public bytes(): u8[] {
        return this.essence().concat(this.signatureScheme.signature);
    }

    public essence(): u8[] {
        let data: u8[] = [1]; // requestKindTagOffLedgerISC
        data = data.concat(this.chainID.toBytes());
        data = data.concat(this.contract.toBytes());
        data = data.concat(this.entryPoint.toBytes());
        data = data.concat(this.params);
        data = data.concat(wasmlib.uint64ToBytes(this.nonce));
        data = data.concat(wasmlib.uint64ToBytes(this.gasBudget));
        const pubKey = wasmlib.bytesFromUint8Array(this.signatureScheme.keyPair.publicKey);
        data = data.concat([pubKey.length as u8]);
        data = data.concat(pubKey);
        //TODO convert to bytes according to isc.Allowance?
        data = data.concat(this.allowance.toBytes());
        return data;
    }

    public ID(): wasmlib.ScRequestID {
        //TODO
        return wasmlib.requestIDFromBytes([]);
    }

    public sign(keyPair: isc.KeyPair): OffLedgerRequest {
        const req = new OffLedgerRequest(this.chainID, this.contract, this.entryPoint, this.params, this.nonce);
        req.signatureScheme = new isc.OffLedgerSignatureScheme(keyPair);
        req.signatureScheme.signature = keyPair.sign(this.essence());
        return req;
    }

    public withAllowance(allowance: wasmlib.ScAssets): void {
        this.allowance = allowance
    }
}

File: packages/wasmvm/wasmclient/src/isc/offledgerrequest.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use crypto::{
    hashes::{blake2b::Blake2b256, Digest},
    signatures::ed25519,
};
use wasmlib::*;

use crate::keypair::KeyPair;

#[derive(Clone)]
pub struct OffLedgerRequest {
    chain_id: ScChainID,
    contract: ScHname,
    entry_point: ScHname,
    params: Vec<u8>,
    signature: OffLedgerSignature,
    nonce: u64,
    allowance: ScAssets,
    gas_budget: u64,
}

#[derive(Clone)]
pub struct OffLedgerSignature {
    public_key: ed25519::PublicKey,
    signature: Vec<u8>,
}

impl OffLedgerSignature {
    pub fn new(public_key: &ed25519::PublicKey) -> Self {
        return OffLedgerSignature {
            public_key: public_key.clone(),
            signature: Vec::new(),
        };
    }
}

impl OffLedgerRequest {
    pub fn new(
        chain_id: &ScChainID,
        contract: &ScHname,
        entry_point: &ScHname,
        params: &[u8],
        nonce: u64,
    ) -> Self {
        return OffLedgerRequest {
            chain_id: chain_id.clone(),
            contract: contract.clone(),
            entry_point: entry_point.clone(),
            params: params.to_vec(),
            signature: OffLedgerSignature::new(&KeyPair::new(&[]).public_key),
            nonce,
            allowance: ScAssets::new(&[]),
            gas_budget: u64::MAX,
        };
    }

    pub fn essence(&self) -> Vec<u8> {
        let mut data: Vec<u8> = vec![1]; // requestKindTagOffLedgerISC
        data.extend(self.chain_id.to_bytes());
        data.extend(self.contract.to_bytes());
        data.extend(self.entry_point.to_bytes());
        data.extend(self.params.clone());
        data.extend(uint64_to_bytes(self.nonce));
        data.extend(uint64_to_bytes(self.gas_budget));
        data.extend(self.allowance.to_bytes());
        return data;
    }

    pub fn id(&self) -> ScRequestID {
        // req id is hash of req bytes with output index zero
        let mut hash = Blake2b256::digest(self.to_bytes()).to_vec();
        hash.push(0);
        hash.push(0);
        return request_id_from_bytes(&hash);
    }

    pub fn sign(&self, key_pair: &KeyPair) -> Self {
        let mut req = OffLedgerRequest::new(
            &self.chain_id,
            &self.contract,
            &self.entry_point,
            &self.params,
            self.nonce,
        );
        req.signature = OffLedgerSignature::new(&key_pair.public_key);
        req.signature.signature = key_pair.sign(&req.essence());
        return req;
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        let mut data = self.essence();
        let public_key = self.signature.public_key.to_bytes();
        data.extend(uint8_to_bytes(public_key.len() as u8));
        data.extend(public_key);
        let signature = &self.signature.signature;
        data.extend(uint16_to_bytes(signature.len() as u16));
        data.extend(signature);
        return data;
    }

    pub fn with_allowance(&mut self, allowance: &ScAssets) -> &Self {
        self.allowance = allowance.clone();
        return self;
    }
}
